File: src/python/plugins/ticket_plugin.py
import requests
from typing import List, Optional
from datetime import datetime
from semantic_kernel.functions import kernel_function

class Ticket:
    def __init__(self, title: str, description: str, assigned_to: Optional[str] = "Matteo Pagani", severity: Optional[str] = "Normal", status: Optional[str] = "Open", created_at: Optional[str] = None, id: Optional[str] = None):
        self.title = title
        self.description = description
        self.assignedTo = assigned_to
        self.severity = severity
        self.status = status
        self.createdAt = created_at or datetime.utcnow().isoformat()
        self.id = id

    @staticmethod
    def from_dict(data: dict):
        return Ticket(
            title=data.get("title"),
            description=data.get("description"),
            assigned_to=data.get("assignedTo"),
            severity=data.get("severity"),
            status=data.get("status"),
            created_at=data.get("createdAt"),
            id=data.get("id")
        )

    def to_dict(self):
        return {
            "title": self.title,
            "description": self.description,
            "assignedTo": self.assignedTo,
            "severity": self.severity,
            "status": self.status,
            "createdAt": self.createdAt
        }

class TicketPlugin:
    BASE_URL = "https://ticket-copilot.azurewebsites.net/api"

    def __init__(self, session: Optional[requests.Session] = None):
        self.session = session or requests.Session()

   
    def name(self) -> str:z
    
    @kernel_function
    def create_ticket(self, title: str, description: str, assigned_to: Optional[str] = "Matteo Pagani", severity: Optional[str] = "Normal", status: Optional[str] = "Open") -> Optional[Ticket]:
        """Create a new ticket with the given details."""
        ticket = Ticket(title, description, assigned_to, severity, status)
        response = self.session.post(f"{self.BASE_URL}/tickets", json=ticket.to_dict())
        if not response.ok:
            print(response.text)
            raise Exception(f"Failed to create ticket: {response.status_code} - {response.text}")
        return Ticket.from_dict(response.json())
    
    @kernel_function
    def get_tickets(self, search: Optional[str] = None, assigned_to: Optional[str] = None, status: Optional[str] = None) -> Optional[List[Ticket]]:
        """Retrieve tickets based on optional search criteria."""
        url = f"{self.BASE_URL}/tickets"
        params = {}
        if search:
            params["search"] = search
        if assigned_to:
            params["assignedTo"] = assigned_to
        if status:
            params["status"] = status
        response = self.session.get(url, params=params)
        if not response.ok:
            raise Exception(f"Failed to get tickets: {response.status_code} - {response.text}")
        return [Ticket.from_dict(item) for item in response.json()]

    @kernel_function
    def delete_ticket(self, id: str) -> None:
        """Delete a ticket by its ID."""
        response = self.session.delete(f"{self.BASE_URL}/tickets/{id}")
        if not response.ok:
            raise Exception(f"Failed to delete ticket: {response.status_code} - {response.text}")

File: src/csharp/OrchestrationSamples/Plugins/TicketPlugin.cs
using System.ComponentModel;
using System.Net.Http.Json;
using Microsoft.SemanticKernel;

namespace OrchestrationSamples.Plugins;

public partial class TicketPlugin
{
    private readonly HttpClient _httpClient;
    private const string BaseUrl = "https://ticket-copilot.azurewebsites.net/api";

    public TicketPlugin(HttpClient? httpClient = null)
    {
        _httpClient = httpClient ?? new HttpClient();
    }

    [KernelFunction, Description("Create a new ticket with the provided details.")]
    // Create a new ticket
    public async Task<Ticket?> CreateTicketAsync([Description("The title of the ticket")]string title, [Description("The description of the ticket")]string description, [Description("The person assigned to the ticket")]string? assignedTo = "Matteo Pagani", [Description("The severity of the ticket")]string? severity = "Normal", [Description("The status of the ticket")] string? status = "Open")
    {
        var ticket = new Ticket
        {
            Title = title,
            Description = description,
            AssignedTo = assignedTo,
            Severity = severity,
            Status = status,
            CreatedAt = DateTimeOffset.UtcNow,
        };

        var response = await _httpClient.PostAsJsonAsync($"{BaseUrl}/tickets", ticket);
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            Console.WriteLine(error);
            throw new Exception($"Failed to create ticket: {response.StatusCode} - {error}");
        }
        return await response.Content.ReadFromJsonAsync<Ticket>();
    }

    // Get tickets (optionally filter by search, assignedTo, status)
    [KernelFunction, Description("Get the list of tickets, optionally filtered by search term, assigned user, and status.")]
    public async Task<List<Ticket>?> GetTicketsAsync([Description("The search term")]string? search = null,[Description("The person assigned to the ticket")] string? assignedTo = null, [Description("The status of the ticket")]string? status = null)
    {
        var url = $"{BaseUrl}/tickets";
        var query = new List<string>();
        if (!string.IsNullOrEmpty(search)) query.Add($"search={Uri.EscapeDataString(search)}");
        if (!string.IsNullOrEmpty(assignedTo)) query.Add($"assignedTo={Uri.EscapeDataString(assignedTo)}");
        if (!string.IsNullOrEmpty(status)) query.Add($"status={Uri.EscapeDataString(status)}");
        if (query.Count > 0) url += "?" + string.Join("&", query);

        var response = await _httpClient.GetAsync(url);
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Failed to get tickets: {response.StatusCode} - {error}");
        }
        return await response.Content.ReadFromJsonAsync<List<Ticket>>();
    }

    // Delete a ticket by id
    public async Task DeleteTicketAsync(string id)
    {
        var response = await _httpClient.DeleteAsync($"{BaseUrl}/tickets/{Uri.EscapeDataString(id)}");
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Failed to delete ticket: {response.StatusCode} - {error}");
        }
    }
}
